doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,214 @@
+keep its table synced.
+
+Ensuring that there is no gap between finishing the initial COPY and starting
+reading from the replication slot is done by synchronizing the database snapshot
+at which the `COPY` is performed with the starting point of the replication
+slot. The simplest way to do this is by using the following query structure:
+
+```sql
+-- Initial sync phase
+BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ
+CREATE_REPLICATION_SLOT <foo> LOGICAL pgoutput USE_SNAPSHOT;
+COPY `table` TO STDOUT;
+COMMIT
+
+-- Streaming phase
+START_REPLICATION SLOT <foo>
+```
+
+By creating the replication slot as the first statement in a transaction and
+setting the `USE_SNAPSHOT` option we ensure that the transaction and the slot
+are at the exact same point. The existence of the slot will prevent any WAL
+segments from being cleaned up while our replica is busy reading the initial
+snapshot.
+
+Once the initial sync is done the transaction is committed to release the
+snapshot and the replica can start streaming the slot normally with no data
+loss.
+
+### Transaction boundaries and timestamps
+
+A key design goal is to maintain the  transaction boundaries from the upstream
+database so that no query on materialize can observe states that never existed
+upstream. At the same time, the data ingested from postgres should be join-able
+with other sources in materialize. This means that all updates must have a
+timestamp that is related to the wall-clock time.
+
+The implementation is greatly simplified by the logical replication protocol
+that guarantees that we'll receive full transactions in commit order. In the
+replication stream each transaction is represented by a BEGIN message, followed
+by one or more row updates, and finally by a COMMIT message.
+
+Each transaction is timestamped with the wall-clock time at the time of the
+BEGIN message. This allows the source implementation to emit rows as it receives
+them and ensure that transaction boundaries are guaranteed.
+
+### Adaptation to multiple tables
+
+If the sources naively applied the mechanism described above for every source
+instance they would [very quickly
+exhaust](https://www.postgresql.org/docs/13/runtime-config-replication.html#GUC-MAX-REPLICATION-SLOTS)
+the available replication slots from upstream and also multiply the bandwidth
+used by several times.
+
+In order to have one independent `SOURCE` object for each upstream table while
+also maintaining a single replication connection between them a new SQL syntax
+will be introduced that allows creating multiple sources at once.
+
+Specifically, a user should be able to do `CREATE SOURCES FROM POSTGRES..` and
+have one `SOURCE` object created for each upstream table. By introducing this
+syntax the SQL planner will have an opportunity to create all the relevant
+sources at once and therefore give them some shared state. The shared state
+could be some form of `Arc<RwLock<Option<T>>>` where the first source instance
```

Comment:
This came up in discussion. This is potentially problematic if it is meant to be shared across workers, but not problematic if it is to be shared across sources but within a worker. I.e., if each of the sources behave a bit like we might want Kafka sources to, sharing the reader for multiple source instances, the no worries and we could mirror that design. If instead the goal is to get some coordination between the workers using `Arc<Mutex<_>>` we should strongly avoid that imo (and instead try and use dataflow to coordinate, if at all possible). One example from the chat was to have one operator that reads single-threaded from postgres (idk) and then distributes its readings to independent operators downstream for each table; this "coordinates" the reading from postgres but still gets the results to distributed source instances.

Reply:
Hm, in the case of postgres sources specifically it is true that the reading of the replication stream is single threaded and also that all the downstream sources are completely dumb and just report the messages received in their collections.

Does what you say mean that we could force all of these sources to run on the same worker and use an `Arc<Mutex<..>>` to spawn the reading task once?